syntax = "proto3";

package cosim;

service Cosim {
  rpc SimulationCall(SimulationAction) returns (SimulationResponse);
  rpc AuthorizeCall(Authorize) returns (AuthorizeResponse);
}

enum State {
  SUCCESS = 0;
  FAILURE = 1;
}

message SimulationAction {
  string cluster_handle_address = 1;
  string job_id = 2;
  string agent_id = 3;
  string external_state_block_id = 4;
  double time = 5;
  oneof request {
    Consume consume = 6;
    Produce produce = 7;
  }
}

message SimulationResponse {
  State state = 1;
  int32 time = 2;
  oneof response {
    ConsumeResponse consume_response = 3;
    ProduceResponse produce_response = 4;
    FailResponse fail_response = 5;
  }
}

message Consume {
  int32 index = 1;
}

message ConsumeResponse {
  double index = 1;
  string value = 2;
}

message Produce {
  int32 index = 1;
  string value = 2;
}

message ProduceResponse {
  int32 index = 1;
}

message FailResponse {
  string message = 1;
}

message Authorize {
  string auth_token = 1;
  string cluster_handle_address = 2;
}

message AuthorizeResponse {
  bool success = 1;
  string session_uuid = 2;
}
